codec/src/test/java/io/netty/handler/codec/FlowControlHandlerTest.java

```diff
@@ -0,0 +1,332 @@
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+
+                // We're turning reading off but we'll continue receiving
+                // channelRead() calls.
+                ctx.channel().config().setAutoRead(false);
+            }
+        };
+
+        EmbeddedChannel channel = new EmbeddedChannel(handler);
+        assertTrue(channel.config().isAutoRead());
+
+        channel.writeInbound("1", "2", "3");
+        assertFalse(channel.config().isAutoRead());
+
+        // channelRead() was called three times even though auto reading was
+        // turned off
+        assertEquals(3, counter.get());
+    }
+
+    /**
+     * This test demonstrates {@link FlowControlHandler}'s behavior with auto
+     * reading turned on from the get-go.
+     */
+    @Test
+    public void testFlowAutoReadingOn() {
+
+        final AtomicInteger counter = new AtomicInteger();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+            }
+        };
+
+        FlowControlHandler flow = new FlowControlHandler();
+
+        EmbeddedChannel channel = new EmbeddedChannel(flow, handler);
+        assertTrue(channel.config().isAutoRead());
+
+        channel.writeInbound("1", "2", "3");
+        assertTrue(channel.config().isAutoRead());
+
+        // All messages were passed thru and flow isn't holding onto any
+        // messages
+        assertEquals(counter.get(), 3);
+        assertTrue(flow.queue().isEmpty());
+    }
+
+    /**
+     * This test demonstrates {@link FlowControlHandler}'s behavior with auto
+     * reading turned off from the get-go.
+     */
+    @Test
+    public void testFlowAutoReadingOff() {
+
+        final AtomicInteger counter = new AtomicInteger();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg)
+                    throws Exception {
+                counter.incrementAndGet();
+            }
+        };
+
+        FlowControlHandler flow = new FlowControlHandler();
+        EmbeddedChannel channel = new EmbeddedChannel(flow, handler);
+
+        channel.config().setAutoRead(false);
+        assertFalse(channel.config().isAutoRead());
+
+        channel.writeInbound("1", "2", "3");
+
+        // We wrote three messages and one made it thru. The reminder of
+        // messages is being held by the FlowControlHandler.
+        //
+        // NOTE: This may seem confusing at first why one message made it
+        // thru but that's because there's one covert read() call in the
+        // EmbeddedChannel class.
```

Comment:
@Scottmitch I don't know if I agree or disagree. I can write a test that involves HttpObjectDecoder to demonstrate the problem and how this class solves it but EmbeddedChannel seems to be adequate for it. `RecvByteBufAllocator` deals with bytes read from the socket. This class deals with messages that were produced by upstream handlers. How these messages came to life doesn't matter. Some messages such as `LastHttpContent#EMPTY_LAST_CONTENT` are not even backed by actual bytes. What matters is that handlers are shoveling messages downstream and this class gives the user a tool to control that flow and builds on top of the existing API.